ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryMonitor.java

```diff
@@ -0,0 +1,215 @@
+  }
+
+  void restartFailedQueries() {
+    ksqlEngine.getPersistentQueries().stream()
+        .forEach(query -> {
+          final QueryId queryId = query.getQueryId();
+          final KafkaStreams.State queryState = query.getState();
+
+          // If the query was restarted previously, check if it needs another restart; or if the
+          // query is now up and running, then remove it from the restart list.
+          if (queriesRetries.containsKey(queryId)) {
+            final RetryEvent retryEvent = queriesRetries.get(queryId);
+            if (ticker.read() > retryEvent.nextRestartTimeMs()) {
+              if (queryState == KafkaStreams.State.ERROR) {
+                // Retry again if it's still in ERROR state
+                retryEvent.restart();
+              } else {
+                // Query is not in ERROR state anymore.
+                queriesRetries.remove(queryId);
+              }
+            }
+          } else if (queryState == KafkaStreams.State.ERROR) {
+            // Restart new query in ERROR state, and add it to the list of retries.
+            final RetryEvent retryEvent = new RetryEvent(
+                ksqlEngine, queryId, BASE_WAITING_TIME_MS, retryBackoffMaxMs, ticker);
+
+            queriesRetries.put(queryId, retryEvent);
+            retryEvent.restart();
+          }
+        });
+  }
+
+  private class Runner implements Runnable {
+    @Override
+    public void run() {
+      LOG.info("KSQL query monitor started.");
+
+      while (!closed) {
+        try {
+          restartFailedQueries();
+        } catch (final Exception e) {
+          LOG.warn("KSQL query monitor found an error attempting to restart failed queries.", e);
+        }
+      }
+    }
+  }
+
+  static class RetryEvent {
+    private final KsqlEngine ksqlEngine;
+    private final QueryId queryId;
+    private final Ticker ticker;
+
+    private int numRetries = 0;
+    private long waitingTimeMs;
+    private long expiryTimeMs;
+    private long retryBackoffMaxMs;
+    private long baseWaitingTimeMs;
+
+    RetryEvent(
+        final KsqlEngine ksqlEngine,
+        final QueryId queryId,
+        final long baseWaitingTimeMs,
+        final long retryBackoffMaxMs,
+        final Ticker ticker
+    ) {
+      this.ksqlEngine = ksqlEngine;
+      this.queryId = queryId;
+      this.ticker = ticker;
+
+      this.baseWaitingTimeMs = baseWaitingTimeMs;
+      this.waitingTimeMs = baseWaitingTimeMs;
+      this.expiryTimeMs = ticker.read() + baseWaitingTimeMs;
+      this.retryBackoffMaxMs = retryBackoffMaxMs;
+    }
+
+    public long nextRestartTimeMs() {
+      return expiryTimeMs;
+    }
+
+    public int getNumRetries() {
+      return numRetries;
+    }
+
+    public void restart() {
+      numRetries++;
+
+      LOG.info("Restarting query {} (attempt #{})", queryId, numRetries);
+
+      // Stop the queryId using the current QueryMetadata
+      ksqlEngine.getPersistentQuery(queryId).ifPresent(q -> q.stop());
```

Comment:
I don't think we want to be calling `stop` here - we should introduce a new method to make sure that `KafkaStreams.cleanUp()` isn't called. If I understand correctly this will actually clean up the state store directory and cause it to restore all of that state from the changelog topic (or expose `doClose` publicly with a better name).

this will also close the state listener, which will remove all of the metrics (which could be problematic for alerting reasons)

Reply:
I added a new method `restart` in the `PersistentQueryMetadata` to handle the close/reset of the KafkaStreams without cleaning it up.